Add 'promoted' access request state

jakule opened the pull request:
This change introduces a new 'promoted' access request state. The state represents that an access request has been promoted to an access list.

Affected code was modified to adjust to the new promoted state and ensure correct system's behavior.


Access request promotion workflow.

A user wants access to a resource. They will log in and create an access request as they do today. 
An access request reviewer may decide to upgrade/promote the request to an access list. Access list gives a user long-term permission to a resource/role (months instead of a few days as access request).

The new state is introduced to make it clearer for the user who created the access request. Access list technically replaces the access request as it gives the same access, but:
* Changing the state to `approve` would give you the same access two times, which can also be confusing from an administration perspective
* Changing the state to `denied` would make it confusing to a user who created the request (my access request was denied, but I have access now). It would also look weird from the audit log perspective.
* Leaving the request in `pending` state doesn't show that any action was taken
* Removing the access request is similar to leaving in `pending`. People will be confused.

Adding the new status seems like the least evil solution as the requester can see that the request was promoted, and the reviewer knows that nothing else will happen with that request (it won't be approved even if a user is added and removed from the access list later).

jakule committed:
This change introduces a new 'promoted' access request state. The state represents that an access request has been promoted to an access list.

Affected code was modified to adjust to the new promoted state and ensure correct system's behavior.

Added a new 'GetAccessRequest' method for internal use to retrieve access request info.
Disallowed direct promotion of access requests. Introduced 'SubmitAccessReviewAllowPromotion' for promotions.
Added 'PromoteAccessRequest' method and updated its usage to restrict direct promotions.
Refactored code for better readability and testing. Renamed some functions, simplified logic, added test helpers.
Introduced 'promoted' state for access requests to handle promotion workflow.
Added 'PromotedAccessListTitle' in 'AccessReview' to track promotion state.

fspmarshall reviewed (commented):
Code itself LGTM, but I'm concerned about the permission model.  As I understand it, anyone with review permissions can submit a review that sets the request to `PROMOTED` state, which is effectively the same as a denied state, but with extra data (the name of the list that the user was promoted into).

This conflicts with the permission model of the review system, as access request RBAC can specify custom denial thresholds, or specify reviewers that can approve but never deny.  If setting a request to the promoted state is functionally equivalent to a denial, then any user with review permissions (even those that would not typically permit any form of denial) can unilaterally deny requests.

Instead of actually updating the request state to `PROMOTED`, what if we just treated promoting review as "information only" (i.e. it informs the user and other reviewers that promotion occurred, but does not interfere with normal approval/denial operations)?  Treating promotion this way would simplify this PR a bit (could remove all elements related to handling requests in a `PROMOTED` state), and ensure that promotion isn't a denial of service vector.

marcoandredinis commented:
You start with an access request, and then you "promote" it to an Access List.
So, you pick the resources within the Access Request/roles and create an Access List with the same set of resources/roles?
Is this how it works?

I might be missing context, so, if you have an issue or some doc you can link to, it would be awesome.

What is the expectation for the plugins?
For example, JIRA plugin works with 3 columns: new, approved, rejected
What column should the Jira issue move to when the access request is promoted?

There's also some plugin development that is entirely managed by each customer.
We should add a clear message in release notes that it might break their workflows if they don't consider this new state.

@EdwardDowling Can you also review this PR? It might have implications on the ServiceNow plugin you are writting.

hugoShaka commented:
>  For example, JIRA plugin works with 3 columns: new, approved, rejected
>  What column should the Jira issue move to when the access request is promoted?

I think this should end up in the approved column because the outcome is that the requester now has access. Teleport should treat this internally as a deny but should display this as an approval.

> This conflicts with the permission model of the review system, as access request RBAC can specify custom denial thresholds, or specify reviewers that can approve but never deny. If setting a request to the promoted state is functionally equivalent to a denial, then any user with review permissions (even those that would not typically permit any form of denial) can unilaterally deny requests.

Based on what I understand, promotion bypasses approval requirements by design. If someone has the right to add you to an access list, they can do it at any time, regardless of you doing an access request. You could just ping the approver on slack, they would add you, and you would have access. The outcome is the same, but the user experience is worse.

From a UX point of view, I think it makes more sense to resolve the access request as soon as the user can now access the resource as no action is needed anymore from approvers. We don't really approve or deny the temporary access request, we just flag it as not needing attention anymore as it was resolved by other means.

hugoShaka reviewed (approved):
LGTM, but we'll need to be extra careful with the plugins. We'll need to send a PR in `teleport-plugins` adding support for this new state, and we might want to do a short test plan that validates that all plugins are dealing with promoted access requests properly once everything is merged.

jakule commented:
> Instead of actually updating the request state to PROMOTED, what if we just treated promoting review as "information only" (i.e. it informs the user and other reviewers that promotion occurred, but does not interfere with normal approval/denial operations)? Treating promotion this way would simplify this PR a bit (could remove all elements related to handling requests in a PROMOTED state), and ensure that promotion isn't a denial of service vector.

@fspmarshall I do understand the concern, but I think that leaving the access request pending makes the whole flow very confusing. Currently the flow is simple `pending` -> `approved/denied`. If we add `promoted` to the mix, then the request may go `peding` -> `promoted` -> `denied`, that will look weird to everyone (your access was promoted, then denied, but you still have access as access list in unrelated to access request in the backend). 

> You start with an access request, and then you "promote" it to an Access List.
So, you pick the resources within the Access Request/roles and create an Access List with the same set of resources/roles?
Is this how it works?
> I might be missing context, so, if you have an issue or some doc you can link to, it would be awesome.

@marcoandredinis As always there is an RFD https://github.com/gravitational/teleport.e/blob/master/rfd/0006e-access-lists.md but I added the gist to the description.

> Based on what I understand, promotion bypasses approval requirements by design. If someone has the right to add you to an access list, they can do it at any time, regardless of you making an access request. You could just ping the approver on Slack, they would add you, and you would have access. The outcome is the same, but the user experience is worse.

That was the main motivation. Also from the audit log perspective, we don't want to log `access request denied` every single time someone uses the access list instead of the access request. It would be probably hard to explain during an audit why so many people were denied access to a resource.

marcoandredinis commented:
I wonder if we could use a subkind instead of adding a new state.
So, when creating the access request you could already say that you want an access list, and then its state would go to approve as usual, and using the same RBAC rules as we use today.

Except that, when approved, an Access List was created instead.

Can an `Access Request` be converted to `AccessRequest:access-list` or it's something that must be decided at creation time?
Not sure.

fspmarshall commented:
> From a UX point of view, I think it makes more sense to resolve the access request as soon as the user can now access the resource, as no approvers need further action. We don't approve or deny the temporary access request, we just flag it as not needing attention anymore as it was resolved by other means.


> I think that leaving the access request pending makes the whole flow very confusing. Currently the flow is simple `pending` -> `approved/denied`. If we add `promoted` to the mix, then the request may go `peding` -> `promoted` -> `denied`, that will look weird to everyone (your access was promoted, then denied, but you still have access as access list in unrelated to access request in the backend).

@hugoShaka @jakule  I don't necessarily disagree about the UX of `pending` -> `promoted` being nice for its simplicity, but the scenario you describe isn't quite what's happening here.  Teleport isn't saying "I see you have permanent access that makes this request redundant, so I'm going to resolve it for you".  Instead, its saying "A user with the ability to write comments on this request *said* this request is redundant... I'm going to take their word of it and resolve it".  Being able to _review_ a request is just commenting permissions. Wether you can actually approve/deny a request depends on the thresholds.  The permission model in the PR lets anyone with review permissions resolve any request.

Speaking analogously, this is like letting anyone with permission to comment on your google doc delete it.  Not good if you care about granular permissions (which, presumably, anyone making non-trivial use of the access request system does care about).

If we want to go through all the work to ensure that teleport is able to detect and auto-resolve redundant access requests, so be it (tho IMO its not worth the effort, and probably would be a bit brittle).  What I don't think we should do is let anyone terminate any request just because they can review it.

Likewise, I think it would be confusing if promotion failed because someone else concurrently approved the request.  Even if the request was an approved this time, it'd be nice if promotion succeeded so that the user could forgo the request next time.  The model just seems cleaner if promotion is an event that can happen orthogonally with approval.

jakule commented:
After having a few more discussions, we decided not to introduce the new `promoted` state but mark the access request as `promoted` when someone submits a review with promotion, but do not change the request status. I'll reopen the PR after fixing the implementation.

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -840,6 +869,18 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *authpb.Requ
+		return nil, trace.BadParameter("state promoted can be only set when promoting to access list")
+	}
+
+	return g.SubmitAccessReviewAllowPromotion(ctx, review)
```
this was confusing to me (b/c we rejected it above), i would rather duplicate a few lines of code and keep it how it was before?

jakule replied:
I'd be very happy to do that, but `SubmitAccessReviewAllowPromotion` was only created as an entry point for the promotion request and it's called from `e`. Otherwise, I've got no idea how can we call the logic in OSS.

kimlisa reviewed (commented):
code lgtm, but don't feel comfortable enough to approve it (i'll punt that to other folks)

jakule commented on api/proto/teleport/accesslist/v1/accesslist_service.proto:
```diff
@@ -176,3 +179,19 @@ message DeleteAllAccessListMembersRequest {
   reserved "access_list";
 }
+
+// PromoteAccessRequestRequest is the request for promoting an access request to an access list.
```
Our linter enforces this so my options are PromoteAccessRequest and PromoteAccessResponse or the above (the name must end with Request/Response). If I feel more creative, I'll change it :see_no_evil:

jakule replied:
I renamed it to `AccessRequestPromote{Request/Response}` to remove the `RequestRequest` from the name

jakule commented on lib/auth/auth_with_roles.go:
```diff
@@ -2669,6 +2669,11 @@ func (a *ServerWithRoles) SetAccessRequestState(ctx context.Context, params type
 		return trace.Wrap(err)
 	}
+
+	if params.State.IsPromoted() {
```
`SetAccessRequestState` changes the state to the provided one, `SubmitAccessReview` creates a review and proposes a new state that is applied to the request if all conditions are met.

fspmarshall replied:
`SetAccessRequestState` used to be the only way to approve access requests.  Originally, approving access requests was a resource-level permission that let you approve _any_ access request unilaterally.  We added the review system later on, with its custom review permissions and thresholds so that individuals could be granted limited review permissions to a subset of requests.

You can think of `SetAccessRequestState` as the "admin" version of approval.  Also, some of our users have automated systems they've written in-house that automatically approve requests via this method.

jakule commented on lib/auth/grpcserver.go:
```diff
@@ -840,6 +869,18 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *authpb.Requ
 func (g *GRPCServer) SubmitAccessReview(ctx context.Context, review *types.AccessReviewSubmission) (*types.AccessRequestV3, error) {
+	// Prevent users from submitting access reviews with the "promoted" state.
+	// Promotion is only allowed by SubmitAccessReviewAllowPromotion API in Ent.
+	if review.Review.ProposedState.IsPromoted() {
```
It doesn't matter in this case. We don't want anyone to submit a promotion through this API regardless of the access list name. If the name is provided (with a different state) the review will be rejected earlier.

fspmarshall commented on lib/auth/grpcserver.go:
```diff
@@ -840,6 +869,18 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *authpb.Requ
 }
 
 func (g *GRPCServer) SubmitAccessReview(ctx context.Context, review *types.AccessReviewSubmission) (*types.AccessRequestV3, error) {
+	// Prevent users from submitting access reviews with the "promoted" state.
+	// Promotion is only allowed by SubmitAccessReviewAllowPromotion API in Ent.
+	if review.Review.ProposedState.IsPromoted() {
+		return nil, trace.BadParameter("state promoted can be only set when promoting to access list")
```
As a general rule, I'd put restrictions like this in `ServerWithRoles` not `GRPCServer`.  There's nothing grpc-specific about this restriction.

jakule replied:
Moved.

fspmarshall commented on lib/auth/grpcserver.go:
```diff
@@ -840,6 +869,18 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *authpb.Requ
+	return g.SubmitAccessReviewAllowPromotion(ctx, review)
+}
+
+// SubmitAccessReviewAllowPromotion is a method created only for internal use by Ent client, and it is not
+// exposed in our GRPC service.
+func (g *GRPCServer) SubmitAccessReviewAllowPromotion(ctx context.Context, review *types.AccessReviewSubmission) (*types.AccessRequestV3, error) {
```
This seems strange to me.  I can't recall us ever needing to create fake methods on `GRPCServer` before.  Fells wrong.  Also, more importantly, this strikes me as a potential risk.  This method looks and feels like a method that is meant to be expose to the network (it's doing authz after all), but it actually circumvents a key access-control mechanism.  Even if it isn't actually exposed, this pattern strikes me as extremely prone to future human error.

jakule replied:
I do agree, but I don't see a good way of sharing code between `e` and `ServerWithRoles`. As a compromise, I extracted the validation logic to `func AuthorizeAccessReviewRequest`. I'm using it in both places. It looks cleaner and less hacky, but it requires the implementation of some part of it in two places.
I think it's an improvement, but I'm also open to better ideas.

fspmarshall commented on lib/services/access_request.go:
```diff
@@ -657,6 +670,9 @@ ProcessReviews:
 			return nil, nil
 		}
+	case lastReview.ProposedState.IsPromoted():
+		// Let the state change. Promoted won't let any access behave like denied.
```
```suggestion
		// Let the state change. Promoted won't grant any access, meaning it is roughly equivalent to denial.
```

jakule committed:
Refactor function and message names for better clarity

The function and message names related to the promotion of an access request to an access list were restructured for better readability and consistency. Names like 'PromoteAccessReqResponse'  have been replaced with more descriptive names such as 'PromoteAccessRequestResponse'. This increases clarity and consistency across the project.

jakule committed:
Remove the hacky GRPC server implementation

jakule committed:
Change method names to be more descriptive

Renamed all instances of 'PromoteAccessRequest' to 'AccessRequestPromote' in multiple files. The new method name provides a more descriptive and clear understanding of the method's function, which improves code readability and maintenance. This change applies to method definitions, comments, and error messages.

jakule commented on lib/auth/auth.go:
```diff
@@ -3443,6 +3443,9 @@ func (a *Server) getValidatedAccessRequest(ctx context.Context, identity tlsca.I
 			return nil, trace.AccessDenied("access request %q has been denied", accessRequestID)
 		}
+		if req.GetState().IsPromoted() {
+			return nil, trace.AccessDenied("access request %q has been promoted. Use access list to access resources.", accessRequestID)
```
We could, but IMO "your request has been promoted" tells you nothing. Mentioned the access list gives you at least something to search from when this happens.

jakule committed:
Refine error message and introduce IsPromoted method

Refined the error message in 'access_request.go' to better indicate that only promoted requests can set the promoted access list title, not just have one. This enhances clarity of error message. Additionally, introduced 'IsPromoted' method in 'access_request.go' file. This method will be useful for quickly checking if a request is in the PROMOTED state.

jakule committed:
Rename variable in SubmitAccessReview method

Renamed the variable "params" to "submission" in the 'SubmitAccessReview' function, in 'auth_with_roles.go' file. The name "submission" provides clearer indictation of its role in submitting access review. This enhances code readability and understandability. No logic changes were made during this update.
